src/compiler/checker.ts

```diff
@@ -23000,6 +22987,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 // method). Simply do a pairwise comparison of the signatures in the two signature lists instead
                 // of the much more expensive N * M comparison matrix we explore below. We erase type parameters
                 // as they are known to always be the same.
+                Debug.assertEqual(sourceSignatures.length, targetSignatures.length);

```

Comment:
They can actually be undefined; lots of anonymous types don't have symbols. More or less it seems like symbols are only around if the user actually wrote the type out or it's named or something. My understanding here is poor.

Reply:
Adding `Debug.assertIsDefined(source.symbol);` here fails 297 tests, FWIW

Reply:
In case I wasn't clear, I didn't mean an assert, just an extra check in the `if` condition that wraps this, since the comment inside it says "We have instantiations of the same anonymous type" and assumes so, but if two instantiated types both have symbol `undefined`, it doesn't seem to me that that's a valid assumption?

Reply:
Yeah, theoretically two symbol-less anonymous symbols could be compared here, but the intent of the code seems to be explicitly that this path is only for things from the same symbol.